### PR TITLE
Bringup OpenAPI

### PR DESCRIPTION
**Pull request by provokateurin**
## Summary

Replaces https://github.com/nextcloud/server/pull/36356, https://github.com/nextcloud/server/pull/36408, https://github.com/nextcloud/server/pull/36517, https://github.com/nextcloud/server/pull/36567 because it was annoying switching between multiple branches/PRs.

## TODO

- [x] https://github.com/nextcloud/server/pull/36515
- [x] https://github.com/nextcloud/server/pull/36665
- [x] https://github.com/nextcloud/server/pull/36667
- [x] https://github.com/nextcloud/server/pull/37390

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/cloud_federation_api/lib/Controller/RequestHandlerController.php [hunk 1]
```diff
@@ -108,45 +118,43 @@ public function __construct($appName,
...
-	 * @param $resourceType ('file', 'calendar',...)
-	 * @return Http\DataResponse|JSONResponse
+	 * @param string|null $ownerDisplayName display name of the user who shared the item
+	 * @param string|null $sharedBy provider specific UID of the user who shared the resource
+	 * @param string|null $sharedByDisplayName display name of the user who shared the resource
+	 * @param array{name: string[], options: array} $protocol e,.g. ['name' => 'webdav', 'options' => ['username' => 'john', 'permissions' => 31]]
+	 * @param string $shareType 'group' or 'user' share
+	 * @param string $resourceType 'file', 'calendar',...
+	 *
+	 * @return JSONResponse201<CloudFederationAddShare>|JSONResponse400<CloudFederationValidationError>|JSONResponse501<CloudFederationError>
```
Comment (nickvergessen) [inline apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:131]: The declared return type
```
'OCP\AppFramework\Http\JSONResponse201<array{recipientDisplayName: string}>|OCP\AppFramework\Http\JSONResponse400<array{message: non-empty-string, validationErrors: array<array-key, array{message: null|string, name: string}>}>|OCP\AppFramework\Http\JSONResponse501<array{message: non-empty-string}>'
```
for OCA\CloudFederationAPI\Controller\RequestHandlerController::addShare is more specific than the inferred return type
```
'OCP\AppFramework\Http\JSONResponse201<array{recipientDisplayName: string}>|OCP\AppFramework\Http\JSONResponse400<array{message: non-empty-string, validationErrors: array<empty, empty>}>|OCP\AppFramework\Http\JSONResponse501<array{message: string}>'
```

For better comparison
```
'OCP\AppFramework\Http\JSONResponse201<array{recipientDisplayName: string}>|OCP\AppFramework\Http\JSONResponse400<array{message: non-empty-string, validationErrors: array<array-key, array{message: null|string, name: string}>}>|OCP\AppFramework\Http\JSONResponse501<array{message: non-empty-string}>'
'OCP\AppFramework\Http\JSONResponse201<array{recipientDisplayName: string}>|OCP\AppFramework\Http\JSONResponse400<array{message: non-empty-string, validationErrors: array<empty, empty>}>|OCP\AppFramework\Http\JSONResponse501<array{message: string}>'
```

**Comment by ChristophWurst** [comment]
(unsubscribing to avoid a notification for every rebase dance. please mention me when there is something to review again)

**Comment by provokateurin** [comment]
CI should finally pass :tada: . The PR isn't 100% ready though and I will split all the changes into multiple PRs.

**Comment by provokateurin** [comment]
All individual PRs for the server apps have been created :tada:

**Pull request closed without merging**